src/main/java/org/jboss/msc/service/ReadableValueImpl.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2018, Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags. See the copyright.txt file in the
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+package org.jboss.msc.service;
+
+import org.jboss.msc.value.Value;
+
+/**
+ * @author <a href="mailto:ropalka@redhat.com">Richard Opalka</a>
+ */
+final class ReadableValueImpl implements Value<Object> {
```

Comment:
I assume that the point of this is to have separate inject and read objects so that users cannot bypass the API with a cast.

Is there a corresponding write side to this?

Is it correct to state that there is generally intended to be (zero or) one instance of this class per service name?

Reply:
Correct @dmlloyd - the main reason is security. If we'd also implement Injector interface here clever MSC API users might guess Value object provided by requires() methods has a potential to implement also Injector interface and they might try to hook into such broken implementation.

Yes, the corresponding write side is WritableValueImpl class provided later (as you noticed).

It isn't correct to state that there is just single instance of this class per service name in this implementation. Current implementation provides this class per ServiceBuilder.
That is two service builders requiring the same service name will obtain two different instances of ReadableValueImpl object. The reason is again security.

I implemented the following security restrictions into this multi value services implementation:

Every service builder receives unique instances of Values and Injectors it requires or provides.
I decided for this because I think it shouldn't be allowed to share Values and Injectors instances
provided by requires() and provides() methods between multiple services.

Reply:
Do you have a scenario in mind where sharing a readable value becomes a security problem?  The value - no matter who has it - will only be available when its corresponding service is available, so all instances of the read side for a given name should be identical and in sync AFAICT; therefore, it would be impossible to distinguish between N values and 1 value.

We could aggressively disable all readable values that are consumed by a service which is not up, but the added memory cost is likely to be a poor tradeoff in this case.  Sharing values is one of the major potential benefits of this approach.

Reply:
I don't have specific scenario in mind where sharing readable value becomes a security problem. Thinking about it now you are right that sharing dependency values of multiple services via single Value instance is one of the major potential benefits of this new approach @dmlloyd . I'll rework ReadableValueImpl this way.